WebVR API Implementation

MortimerGoro opened the pull request:
<!-- Please describe your changes on the following line: -->

WebVR API Implementation with HTC Vive support on Windows. The current implementations only enables the WebVR support on Windows. In other platforms the API is available on JavaScript but navigator.vr.getDisplays() returns an empty array. This will change when we add support for more VR providers and platforms ;) 

Info about the architecture: 
https://blog.mozvr.com/webvr-servo-architecture-and-latency-optimizations/
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

Proprietary openvr.dll must be copied next to servo.exe in order to test on HTC Vive (https://github.com/ValveSoftware/openvr/tree/master/bin/win64) I have added some of the official WebVR samples for testing. Switch on your headset and run:

mach run tests/html/webvr/room-scale.html

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14618)
<!-- Reviewable:end -->

MortimerGoro committed:
WebVR API Implementation, r=larsbergstrom

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/lib.rs, components/constellation/pipeline.rs, components/constellation/Cargo.toml, components/constellation/constellation.rs
 * @kichjang: components/script/dom/webidls/VRDisplayCapabilities.webidl, components/script/dom/navigator.rs, components/script/dom/vrdisplay.rs, components/script/dom/webidls/VRFrameData.webidl, components/script/dom/mod.rs, components/script/dom/vr.rs, components/script/dom/vrdisplaycapabilities.rs, components/script/dom/vrstageparameters.rs, components/script/dom/vrdisplayevent.rs, components/script/dom/window.rs, components/script/dom/webidls/VREyeParameters.webidl, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script/lib.rs, components/script/dom/webidls/Window.webidl, components/script/dom/vreyeparameters.rs, components/script/dom/webidls/VRStageParameters.webidl, components/script/dom/vrframedata.rs, components/script/dom/macros.rs, components/script/dom/vrpose.rs, components/script/dom/webidls/Navigator.webidl, components/script/dom/webidls/VR.webidl, components/script/dom/webidls/VRPose.webidl, components/script/Cargo.toml, components/script/dom/webidls/VRDisplay.webidl, components/script/dom/webidls/VRDisplayEvent.webidl, components/script/script_thread.rs, components/script/script_runtime.rs, components/script/dom/vrfieldofview.rs, components/script/dom/webidls/VRFieldOfView.webidl, components/script/dom/bindings/conversions.rs, components/script/dom/webidls/VRLayer.webidl, components/script_traits/Cargo.toml, components/script_traits/Cargo.toml
 * @fitzgen: components/script/dom/webidls/VRDisplayCapabilities.webidl, components/script/dom/navigator.rs, components/script/dom/vrdisplay.rs, components/script/dom/webidls/VRFrameData.webidl, components/script/dom/mod.rs, components/script/dom/vr.rs, components/script/dom/vrdisplaycapabilities.rs, components/script/dom/vrstageparameters.rs, components/script/dom/vrdisplayevent.rs, components/script/dom/window.rs, components/script/dom/webidls/VREyeParameters.webidl, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script/lib.rs, components/script/dom/webidls/Window.webidl, components/script/dom/vreyeparameters.rs, components/script/dom/webidls/VRStageParameters.webidl, components/script/dom/vrframedata.rs, components/script/dom/macros.rs, components/script/dom/vrpose.rs, components/script/dom/webidls/Navigator.webidl, components/script/dom/webidls/VR.webidl, components/script/dom/webidls/VRPose.webidl, components/profile/time.rs, components/script/Cargo.toml, components/script/dom/webidls/VRDisplay.webidl, components/profile_traits/time.rs, components/profile_traits/time.rs, components/script/dom/webidls/VRDisplayEvent.webidl, components/script/script_thread.rs, components/script/script_runtime.rs, components/script/dom/vrfieldofview.rs, components/script/dom/webidls/VRFieldOfView.webidl, components/script/dom/bindings/conversions.rs, components/script/dom/webidls/VRLayer.webidl, components/script_traits/Cargo.toml, components/script_traits/Cargo.toml
 * @emilio: components/canvas/webgl_paint_thread.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

emilio commented on components/script/script_thread.rs:
```diff
@@ -2197,6 +2207,13 @@ impl ScriptThread {
     }
 
+    fn handle_webvr_event(&self, pipeline_id: PipelineId, event: WebVREventMsg) {
+        if let Some(window) = self.documents.borrow().find_window(pipeline_id) {
```
Use a different variable for this, see #14589.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+        sender
+    }
+
+    #[allow(unsafe_code)]
```
there doesn't seem to be unsafe code here?

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+
+    fn poll_events(&mut self, sender: IpcSender<bool>) {
+        let events = self.service.poll_events();
+        if events.len() > 0 {
```
`if !events.is_empty()`

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    fn schedule_poll_events(&mut self) {
+        if self.service.is_initialized() && !self.polling_events {
```
nit: Probably nicer to read as:

```rust
if !self.service.is_initialized() || self.polling_events {
    return;
}

// ...
```

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+                        // WebVR Thread asked to unschedule this thread
+                        break;
+                    }
+                    thread::sleep(time::Duration::from_millis(500));
```
This seems quite arbitrary. I understand why it's this way, but probably worth moving the `500` to a constant.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+// This also avoids "JS DDoS" attacks: A Second JSContext doing a lot of calls.
+// while the main one is presenting and demands both high framerate and low latency.
+
+pub struct WebVRCompositor(*mut VRDevice);
```
huh, this is going to need a better comment explaining how it's used and why it's safe and not racy.

emilio replied:
Also, would be helpful to know when this `VRDevice` can be null or not, and how does is this `VRDevice` freed, since I don't see any `Drop` impl.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+                        (*compositor.0).sync_poses();
+                        (*compositor.0).synced_frame_data(near, far).to_bytes()
+                    };
+                    let _result = sender.send(Ok(pose));
```
The patterns `let _ = sender.send(..)` or `drop(sender.send())` seem more used across the codebase.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+impl WebVRCompositorHandler {
+    #[allow(unsafe_code)]
+    fn create_compositor(&mut self, device_id: webrender_traits::VRCompositorId) {
+        if let Some(ref sender) = self.webvr_thread_sender {
```
Probably worth:

```rust
let sender = match self.webvr_thread_sender {
    Some(ref s) => s,
    None => return,
};
```

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,332 @@
+        }
+    }
+
+    pub fn set_webvr_thread_sender(&mut self, sender: IpcSender<WebVRMsg>) {
```
Worth pointing here with a comment that this is done only a per-platform basis on initialization.

emilio reviewed (commented):
Just some mostly-superficial comments, since @larsbergstrom has kindly offered to take over the review.

I guess adding automatic tests for this is not truly feasible? Is there a plan to test this stuff?

asajeffrey commented on components/constellation/pipeline.rs:
```diff
@@ -119,6 +120,8 @@ pub struct InitialPipelineState {
     /// Whether this pipeline is considered private.
     pub is_private: bool,
+    /// A channel to the webvr thread.
+    pub webvr_thread: Option<IpcSender<WebVRMsg>>,
```
When is this used?

MortimerGoro replied:
It's used to create the ScriptThreadFactory. When a script thread is created it gets a cloned handle to the webvr_thread sender. This sender  is used by DOMObjects to comunicate with the WebVRThread and send messages like: getDisplay, requestPresent, etc.

MortimerGoro commented:
@emilio WRT automated testing it could be done with mocked devices, but it requires more work: like setting up predefined pose animations and comparing images submitted to the display. I've a initial implementation of a MockVRDevice and @dmarcos was working on a mock openvr_api.dll. @dmarcos is there any plan for official WebVR conformance tests? Maybe you have talked about that in the latest W3C workshop/meetings.

dmarcos commented:
@MortimerGoro @emilio No official plan for conformance tests. There's a face to face meeting scheduled in January that will touch the subject (among other things). I think in the short we can use your MockVRDevice. In the mid term and for cross browser testing will be better to implement a mock OpenVR driver (instead of the full OpenVR API implementation I had started). In the long term we will probably move to the driver standard that is supposed to come from the khronos vr effort.

Ms2ger commented:
Seems like this should all be behind a pref.

MortimerGoro commented:
@Ms2ger Sounds good, I'll add a new pref to https://github.com/servo/servo/blob/master/components/config/opts.rs that decides if the WebVR API is enabled.

jdm commented:
@MortimerGoro It belongs in resources/prefs.json and annotated in the WebIDL files using [Pref=] instead of a command-line option.

MortimerGoro commented:
@jdm Thanks for the info, never had played with that json before ;)

nox commented on components/script/dom/bindings/conversions.rs:
```diff
@@ -463,6 +466,9 @@ pub unsafe trait ArrayBufferViewContents: Clone {
     fn is_type_compatible(ty: Type) -> bool;
+
+    /// Creates a typed array
+    fn new(cx: *mut JSContext, num: u32) -> *mut JSObject;
```
This should be unsafe.

nox commented on components/script/dom/bindings/conversions.rs:
```diff
@@ -595,3 +633,25 @@ pub unsafe fn is_array_like(cx: *mut JSContext, value: HandleValue) -> bool {
 }
+
+/// Creates a typed JS array from a Rust slice
+pub fn slice_to_array_buffer_view<T>(cx: *mut JSContext, data: &[T]) -> *mut JSObject
```
This should be unsafe.

nox commented on components/script/dom/navigator.rs:
```diff
@@ -114,4 +118,18 @@ impl NavigatorMethods for Navigator {
+
+impl Navigator {
+     pub fn handle_webvr_event(&self, event: WebVREventMsg) {
+         if let Some(vr) = self.vr.get() {
```
How can this be called with `self.vr` being `None`?

nox commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+            return promise;
+        }
+
+        if let Some(wevbr_sender) = self.webvr_thread() {
```
Spelling of `webvr_sender` is wrong, and it is weird to call the variable `sender` when the method says `thread`.

nox commented on components/script/dom/vrdisplaycapabilities.rs:
```diff
@@ -0,0 +1,69 @@
+    capabilities: DOMRefCell<WebVRDisplayCapabilities>
+}
+
+// Wrappers required to include WebVR structs in a DOM struct
```
Why?

nox commented on components/script/dom/vrdisplayevent.rs:
```diff
@@ -0,0 +1,116 @@
+
+impl VRDisplayEvent {
+    fn new_inherited(display: &VRDisplay,
+                     reason: &Option<VRDisplayEventReason>)
```
Why a ref? Pass `Option<VRDisplayEventReason>`.

nox commented on components/script/dom/vrdisplayevent.rs:
```diff
@@ -0,0 +1,116 @@
+                          display: &VRDisplay,
+                          event: &webvr::VRDisplayEvent)
+                          -> Root<VRDisplayEvent> {
+        let (name, reason) = match event {
```
Nit: `match *event`.

nox commented on components/script/dom/vreyeparameters.rs:
```diff
@@ -0,0 +1,74 @@
+    fov: JS<VRFieldOfView>,
+}
+
+// Wrappers required to include WebVR structs in a DOM struct
```
Why?

MortimerGoro replied:
I added the wrappers because Rust didn't allow to implement no_jsmanaged_fields macro or JSTraceable for structs defined in a different crate. I have done a test with the latest rustc compiler used in Servo and it seems to work now. I'll get rid of them now ;)

nox commented on components/script/dom/vrdisplayevent.rs:
```diff
@@ -0,0 +1,116 @@
+        let (name, reason) = match event {
+            &webvr::VRDisplayEvent::Connect(_) => ("displayconnect", None),
+            &webvr::VRDisplayEvent::Disconnect(_) => ("displaydisconnect", None),
+            &webvr::VRDisplayEvent::Activate(_, ref reason) => ("activate", Some(reason)),
```
Why `ref`? Seems to me like VRDisplayEventReason should be `Clone` and `Copy`.

nox commented on components/script/dom/vreyeparameters.rs:
```diff
@@ -0,0 +1,74 @@
+
+impl VREyeParameters {
+    #[allow(unrooted_must_root)]
+    fn new_inherited(parameters: &webvr::VREyeParameters, global: &GlobalScope) -> VREyeParameters {
```
Why a reference?

nox reviewed (commented):
Started looking at it a bit.

cvan commented on components/script/dom/webidls/VRDisplay.webidl:
```diff
@@ -0,0 +1,131 @@
+   * any of the passed layers the promise is rejected.
+   * If the source of any of the layers is not present (null), the promise is rejected.
+   */
+  Promise<void> requestPresent(sequence<VRLayer> layers);
```
FYI, this is fine for now, but see this discussion:

* https://github.com/w3c/webvr/pull/122
* https://github.com/w3c/webvr/pull/143/files

cvan commented on components/script/dom/webidls/VRDisplay.webidl:
```diff
@@ -0,0 +1,131 @@
+  /**
+   * Get the layers currently being presented.
+   */
+  //sequence<VRLayer> getLayers();
```
can you file a new issue to support this?

cvan commented on components/script/dom/webidls/VRDisplayEvent.webidl:
```diff
@@ -0,0 +1,22 @@
+enum VRDisplayEventReason {
+  "navigation",
+  "mounted",
+  "unmounted",
```
can you also support `requested` (see [this spec commit](https://github.com/w3c/webvr/commit/e4c839)?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+                WebVRMsg::PollEvents(sender) => {
+                    self.poll_events(sender);
+                },
+                WebVRMsg::GetVRDisplays(sender) => {
```
can we call this just `GetDisplays`?

cvan commented on tests/html/webvr/simple-mirroring.html:
```diff
@@ -0,0 +1,262 @@
+        // canvas backbuffer need to be made. As a result, we ONLY want to set
+        // that if we know the VRDisplay has an external display, which is why
+        // we defer WebGL initialization until after we've gotten results back
+        // from navigator.getVRDisplays and know which device we'll be
```
will want to update this comment (and others) to reflect your `navigator.vr` change

cvan commented on components/script/dom/webidls/VRLayer.webidl:
```diff
@@ -0,0 +1,13 @@
+
+// https://w3c.github.io/webvr/#interface-vrlayer
+
+//typedef (HTMLCanvasElement or OffscreenCanvas) VRSource;
```
is there any harm in introducing the `VRSource` without support in Servo for `OffscreenCanvas` yet? (we have the API already shipped in Firefox, albeit behind a flag.) filed servo issue #14627 to track the implementation of it.

emilio replied:
Thanks for filling that :)

It's probably fine to add `typedef HTMLCanvasElement VRSource` for now, but I don't think it's worth it.

cvan commented on components/script/dom/webidls/VRLayer.webidl:
```diff
@@ -0,0 +1,13 @@
+
+dictionary VRLayer {
+  HTMLCanvasElement source;
+  sequence<float> leftBounds;
```
sure these shouldn't be `readonly attribute` with defaults of `[]`?

do we not want [`VRLayerInit`](https://w3c.github.io/webvr/#ref-for-dictdef-vrlayerinit-3), per the spec?

MortimerGoro replied:
VRLayerInit is not implemented because some problems with the python autogeneration tool (readonly sequence<T> attributes). I'll file a issue about this

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+                    self.handle_exit_present(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::CreateCompositor(device_id) => {
+                    self.handle_create_compositor(device_id);
```
when does the compositor get disengaged if not needed for browsing context (non-VR) anymore?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+// VRDevices pointers are guaranteed to be valid memory:
+//    * VRDevices are owned by the VRDeviceManager which lives in the WebVRThread.
+//    * WebVRCompositorHandler is stopped automatically when a JS tab is closed or the whole browser is closed.
+//    * WebVRThread and it's VRDevices are destroyed after all tabs are dropped and the browser is about to exit.
```
`it's` -> `its`

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+// Defines the polling interval time in ms for VR Events such as VRDevice connected, disconnected, etc.
+const EVENT_POLLING_INTERVAL: u64 = 500;
+
+// WebVRThread owns native VRDevices, handles their life cycle inside Servo and
```
`VRDevice` -> `VRDisplay`

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    fn access_check(&self, pipeline: PipelineId, device_id: u64) -> Result<&VRDevicePtr, &'static str> {
```
can we change `VRDevicePtr` -> `VRDisplayTr`?

cvan replied:
`device_id` -> `display_id`?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+
+    fn access_check(&self, pipeline: PipelineId, device_id: u64) -> Result<&VRDevicePtr, &'static str> {
+        if *self.presenting.get(&device_id).unwrap_or(&pipeline) != pipeline {
+            return Err("Device owned by another context");
```
`Device` -> `VR Display`

`owned by another context` -> can we be more descriptive?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    fn handle_request_present(&mut self,
+                         pipeline: PipelineId,
```
consistent indentation?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+// Multithreading won't be a problem because:
+//    * Thanks to the security rules implemented in the WebVRThread, when a VRDisplay is in a presenting loop
+//      no other JSContext is granted access to the VRDisplay. So really there aren’t multithreading race conditions.
+//    * VRDevice implementations are designed to allow calling compositor functions
```
can you change all instances of `VRDevice` to `VRDisplay`?

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+                self.compositors.insert(device_id, device);
+            },
+            None => {
+                error!("VRDevice not found when creating a new VRCompositor");
```
`VRDisplay` -> `VRDisplay` (or `VR Display`, if want to be consistent everywhere w/ that usage)

cvan commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+        };
+    }
+
+    // This is done only a per-platform basis on initialization.
```
`done only` -> `done on only`

cvan commented on components/vr_traits/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+serde_derive = "0.8"
+rust-webvr = {version = "0.1", features = ["serde-serialization"] }
+
+
```
feel free to remove these two superfluous newlines, but keep a single one at EOF

cvan commented on components/vr_traits/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+msg = {path = "../msg"}
+serde = "0.8"
+serde_derive = "0.8"
+rust-webvr = {version = "0.1", features = ["serde-serialization"] }
```
can remove the space before the closing curly brace at the end: ` }`

cvan commented on components/vr_traits/webvr_traits.rs:
```diff
@@ -0,0 +1,24 @@
+
+pub type WebVRResult<T> = Result<T, String>;
+
+// Messages from Script Thread to WebVR thread.
```
`Script Thread` -> `Script thread`

cvan commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+
+        if let Some(wevbr_sender) = self.webvr_thread() {
+            let (sender, receiver) = ipc::channel().unwrap();
+            wevbr_sender.send(WebVRMsg::GetVRDisplays(sender)).unwrap();
```
also: `GetDisplays`

cvan commented on components/vr_traits/webvr_traits.rs:
```diff
@@ -0,0 +1,24 @@
+    RegisterContext(PipelineId),
+    UnregisterContext(PipelineId),
+    PollEvents(IpcSender<bool>),
+    GetVRDisplays(IpcSender<WebVRResult<Vec<VRDisplayData>>>),
```
same here: `GetDisplays`

cvan reviewed (commented):
just feedback from having worked on the spec and helped Kip with Gecko implementation + bugs in the past. I'm not familiar with Rust, so take my webdev ergonomics POV with a grain of salt.

I'd like to iron out the event issues, but all means if the reviewers here okay this PR, feel free to proceed.

truly great work - this is really inspiring to see.

jdm commented on components/script/dom/vrstageparameters.rs:
```diff
@@ -0,0 +1,74 @@
+    #[allow(unsafe_code)]
+    // https://w3c.github.io/webvr/#dom-vrstageparameters-sittingtostandingtransform
+    unsafe fn SittingToStandingTransform(&self, _cx: *mut JSContext) -> NonZero<*mut JSObject> {
+        NonZero::new(self.transform.get())
```
Nope.

emilio commented on components/canvas/canvas_paint_thread.rs:
```diff
@@ -209,6 +209,7 @@ impl<'a> CanvasPaintThread<'a> {
                         }
                     }
                     CanvasMsg::WebGL(_) => panic!("Wrong message sent to Canvas2D thread"),
+                    CanvasMsg::WebVR(_) => panic!("Wrong message sent to Canvas2D thread"),
```
It'd be nice to differentiate these error messages now.

emilio commented on components/canvas/webgl_paint_thread.rs:
```diff
@@ -144,6 +144,18 @@ impl WebGLPaintThread {
+            WebGLPaintTaskData::WebRender(ref api, id) => {
+                api.send_vr_compositor_command(id, message);
+            }
+            WebGLPaintTaskData::Readback(_, _, _) => {
```
nit: You can use `WebGLPaintTaskData::Readback(..)` if you want.

emilio commented on components/constellation/constellation.rs:
```diff
@@ -872,6 +879,13 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
                 self.handle_log_entry(top_level_frame_id, thread_name, entry);
             }
+            FromCompositorMsg::SetWebVRThread(webvr_thread) => {
+                self.webvr_thread = Some(webvr_thread)
```
Probably is worth sanity-checking with something like: `assert!(self.webvr_thread.is_none())`

emilio commented on components/constellation/constellation.rs:
```diff
@@ -872,6 +879,13 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+                self.webvr_thread = Some(webvr_thread)
+            }
+            FromCompositorMsg::WebVREvent(pipeline_ids, event) => {
+                debug!("constellation got webvr event message");
```
Also, if we go this spell-checking mode, you may want to capitalize `WebVR` :P

emilio commented on components/script/dom/navigator.rs:
```diff
@@ -114,4 +118,18 @@ impl NavigatorMethods for Navigator {
+    fn Vr(&self) -> Root<VR> {
+        self.vr.or_init(|| VR::new(&self.global()))
+    }
+
```
nit: Extra newline.

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+
+    // https://w3c.github.io/webvr/#interface-navigator
+    fn VrEnabled(&self) -> bool {
+        // TODO: check iframe
```
Yes, please fill a tracking issue with all the remaining APIs and pending changes. Feel free to cc me on that.

MortimerGoro replied:
I'll remove it because vrEnabled and allowvr are about to be removed from spec: https://github.com/w3c/webvr/issues/25

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    pub fn handle_webvr_event(&self, event: WebVREventMsg) {
+        let event = match event {
```
I think you can do:

```rust
let WebVREventMsg::DisplayEvent(event) = event;
```

That should also error out when you add more variants in the future.

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+            WebVREventMsg::DisplayEvent(display_event) => display_event
+        };
+
+        match &event {
```
nit: You can do `match event` directly, and remove all those ampersands.

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+                    self.notify_event(&display, &event);
+                }
+            },
+            &webvr::VRDisplayEvent::Activate(ref display, _) |
```
Oh I see, this is because you borrow event afterwards, right?

Then you can do something like:

```rust
ref event @ webvr::VRDisplayEvent::Deactivate(ref display) |
// ...
```

Though I'm not sure that'd be a strict improvement.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+            };
+        }
+
+        false
```
Can this happen? How can you get a `VRDisplay` without WebVR enabled?

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+            &webvr::VRDisplayEvent::PresentChange(ref display, _) |
+            &webvr::VRDisplayEvent::Change(ref display) => {
+                self.sync_display(&display);
+                if let Some(display) = self.find_display(display.display_id) {
```
Is it expected to find no display here? If it isn't, please `expect`, or at least spit an error.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-resetpose
+    fn ResetPose(&self) -> () {
+        if let Some(wevbr_sender) = self.webvr_thread() {
```
Same for most if not all of these methods. Seems like `webvr_thread` should be infallible, and then you should be able to dedent all this code.

MortimerGoro replied:
It shouldn't happen unless we want to allow something like stopping/disabling WebVR when Servo is already running (without requiring to restart it).

What's the guideline for this kind of cases on Servo, a panic or a console error?

emilio replied:
I think a panic is acceptable. Something like:

```rust
    fn webvr_thread(&self) -> Option<IpcSender<WebVRMsg>> {
        self.global().as_window().webvr_thread().expect("Shouldn't arrive here with WebVR disabled")
    }
```

asajeffrey replied:
Either a panic! or a warn! depending on whether you want to submit a crash report or not. warn! gets included in any later crash report, info! and below does not.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+        // such as compositing WebGL and DOM elements together.
+        // That functionality is not allowed by this revision of the spec.
+        if layers.len() != 1 {
+            let msg = "The number of layers must be 1".to_string();
```
You shouldn't need to convert to string, here and everywhere else in the file, `str` implements `ToJSValConvertible`.

MortimerGoro replied:
I had to use string because of this: 
```
283 |             promise.reject_native(promise.global().get_cx(), msg);
    |                     ^^^^^^^^^^^^^ the trait `core::marker::Sized` is not implemented for `str`
```

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+        }
+
+        // Parse and validate received VRLayer
+        let layer = validate_layer(self.global().get_cx(), &layers[0]);
```
Use `match` instead of `if let Err` and then `unwrap`.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+        } else {
+            self.stage_params.set(None);
+        }
+
```
nit: Extra newline.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+        let near_init = self.depth_near.get();
+        let far_init = self.depth_far.get();
+
+        thread::Builder::new().name("WebVR_RAF".into()).spawn(move || {
```
Hmm... We create this once per `VRDisplay`? Can we do better?

MortimerGoro replied:
It's must be one per display, because it waits VSync for the specific display. I'll improve the documentation ;)

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+            if values.len() == 0 {
+                return Ok(())
+            }
+            if values.len() > 4 {
```
Seems like this doesn't enforce that. Should this be changed to `if values.len() != 4`? Otherwise if I give you three values that's a panic or an out of bounds read.

emilio commented on components/script/dom/vrdisplay.rs:
```diff
@@ -0,0 +1,618 @@
+        match self.frame_data_status.get() {
+            VRFrameDataStatus::Synced => {
+                end_sender.send(Ok((self.depth_near.get(), self.depth_far.get()))).unwrap();
+            },
```
Can you use explicit matching here? It's easier to see how this finishes the RAF thread.

emilio replied:
In general a few notes about what's the lifetime of the display's request animation frame thread would be nice. I guess one thread per display would be acceptable, but a few more docs on how is that managed would help reviewers (blink), to not have to figure it on their own.

emilio commented on components/script/dom/vrdisplaycapabilities.rs:
```diff
@@ -0,0 +1,69 @@
+
+    // https://w3c.github.io/webvr/#dom-vrdisplaycapabilities-maxlayers
+    fn MaxLayers(&self) -> u32 {
+        if self.capabilities.borrow().0.can_present {
```
nit: This may be more legible like:

```rust
if self.CanPresent() { 1 } else { 0 }
```

emilio commented on components/script/dom/vrframedata.rs:
```diff
@@ -0,0 +1,119 @@
+        framedata.right_view.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
+
+        reflect_dom_object(box framedata,
+                    global,
```
nit: indentation

emilio commented on components/script/dom/vrpose.rs:
```diff
@@ -0,0 +1,132 @@
+}
+
+#[allow(unsafe_code)]
+fn update_typed_array(cx: *mut JSContext,
```
probably should be called `update_or_create_typed_array`.

emilio replied:
And should be unsafe.

emilio commented on components/script/dom/vrpose.rs:
```diff
@@ -0,0 +1,132 @@
+
+    pub fn new(global: &GlobalScope, pose: &webvr::VRPose) -> Root<VRPose> {
+        let root = reflect_dom_object(box VRPose::new_inherited(),
+                           global,
```
nit: Indentation

emilio commented on components/vr/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+default_features = false
+features = ["serde_derive"]
+
+
```
Yeah, please delete these trailing newlines.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+use vr_traits::webvr::*;
+use webrender_traits;
+
+// Defines the polling interval time in ms for VR Events such as VRDevice connected, disconnected, etc.
```
Also, I think you can use doc comments for these.

emilio commented on components/vr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+//    * WebVRCompositorHandler is stopped automatically when a JS tab is closed or the whole browser is closed.
+//    * WebVRThread and it's VRDevices are destroyed after all tabs are dropped and the browser is about to exit.
+//      WebVRThread is closed using the Exit message.
+
```
Doc comment above, then remove this newline :)

emilio reviewed (commented):
I took a closer look, here are a few comments :)

MortimerGoro commented on components/script/dom/macros.rs:
```diff
@@ -495,6 +495,13 @@ macro_rules! window_event_handlers(
         event_handler!(unhandledrejection, GetOnunhandledrejection,
                        SetOnunhandledrejection);
         event_handler!(unload, GetOnunload, SetOnunload);
+        event_handler!(vrdisplayconnect, GetOnvrdisplayconnect, SetOnvrdisplayconnect);
```
Yes, I'll remove these ones. I started implementing 1.1 API but moved to 1.2. These events are already moved to VRDisplay instances (in the implementation and samples)

MortimerGoro commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,173 @@
+        self.displays.borrow()
+                     .iter()
+                     .find(|d| d.get_display_id() == display_id)
+                     .map(|d| Root::from_ref(&**d))
```
Yes, because it's a implicit return

MortimerGoro commented on components/servo/lib.rs:
```diff
@@ -296,6 +301,16 @@ fn create_constellation(user_agent: Cow<'static, str>,
                         layout_thread::LayoutThread,
                         script::script_thread::ScriptThread>::start(initial_state);
 
+    if cfg!(target_os = "windows") {
```
It's just because HTC Vive only works on Windows now (no other SDKs implemented yet). I'm going to add a preference for this instead of platform filtering.

MortimerGoro commented on components/constellation/constellation.rs:
```diff
@@ -1181,6 +1195,13 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
 
+        if let Some(chan) = self.webvr_thread.as_ref() {
+            debug!("Exiting WebVR thread.");
```
I included it inside the `if` because WebVR can be disabled and it didn't make sense to show that message. I checked and the same pattern is used in ""Exiting devtools" too ;)

MortimerGoro commented on components/script/dom/webidls/VR.webidl:
```diff
@@ -0,0 +1,10 @@
+interface VR: EventTarget {
+  Promise<sequence<VRDisplay>> getDisplays();
+  //readonly attribute FrozenArray<VRDisplay> activeVRDisplays;
+  readonly attribute boolean vrEnabled;
```
IMO getAvailability or isEnabled() is a useful API to check VR availability because it doesn't load VRDisplays. I mean that you can check availability using the getDisplays promise but in every browser that probably wakes up the lazy loaded WebVR thread, resources, and more. In a webpage you may want to only know availability without loading all those resources.

MortimerGoro replied:
Probably we need to wait for this: https://github.com/WICG/feature-policy

MortimerGoro commented:
@nox @jdm @emilio @larsbergstrom @cvan @Ms2ger Thanks for the feedback. I have updated the PR:
- Everything is behind the dom.webvr.enabled preference now
- Used expect instead of If let Some() in places where the webvr_thread is guaranteed to be enabled
- Renamed VRDevice to VRDisplay
- Fixed nits
- There are two trivial TODOs (getLayers and activeDisplays functions) because of a issue with the python WebIDL code autogeneration tool: it crashes when using `readonly attribute sequence<T>` pattern. I'll fill a issue about this.

larsbergstrom commented on resources/prefs.json:
```diff
@@ -7,6 +7,7 @@
   "dom.serviceworker.timeout_seconds": 60,
   "dom.testable_crash.enabled": false,
   "dom.testbinding.enabled": false,
+  "dom.webvr.enabled": true,
```
This should be `false` by default

larsbergstrom commented on components/webvr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+        if !events.is_empty() {
+            let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
+            for event in events {
+                let event = WebVREventMsg::DisplayEvent(event);
```
There's a bit of duplicated code here with `notify_events` except for the lifting of the `pipeline_ids`. Can this be shared?

larsbergstrom commented on components/webvr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+
+    fn poll_events(&mut self, sender: IpcSender<bool>) {
+        let events = self.service.poll_events();
+        if !events.is_empty() {
```
Can more events come in while you are handling the current set? Elsewhere in Servo event polling threads, we use a pattern that is more like `while (events = poll_events() && !events.is_empty()) { ... }` to avoid falling into the polling delay interval during periods of rapid interactions / event floods.

larsbergstrom commented on components/webvr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+/// by flooding the WebVRThread with messages while the main JavaScript tab is presenting to the headset.
+/// Multithreading won't be a problem because:
+///    * Thanks to the security rules implemented in the WebVRThread, when a VRDisplay is in a presenting loop
+///      no other JSContext is granted access to the VRDisplay. So really there aren’t multithreading race conditions.
```
I assume that this is the `access_check` method above? If so, please make a comment about its importance near that method so people don't accidentally remove / "optimize" it away :-)

MortimerGoro replied:
Yes, I added more comments ;)

larsbergstrom commented on components/webvr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+
+/// WebVRThread owns native VRDisplays, handles their life cycle inside Servo and
+/// acts a doorman for untrusted VR requests from DOM Objects.
+/// It waits for VR Commands from DOM objects and handles them in its trusted thread.
```
Could you add some more info on the design here? There are some long-running and short-lived threads plus communication back with the constellation and the lifetimes and expectations for some of the parts was not totally clear without a pretty deep reading. e.g., which threads have a lifetime tied to a pipeline vs. a single GetDisplays request vs. etc.

larsbergstrom commented on components/webvr/webvr_thread.rs:
```diff
@@ -0,0 +1,349 @@
+use webvr_traits::webvr::*;
+
+/// Defines the polling interval time in ms for VR Events such as VRDisplay connected, disconnected, etc.
+const EVENT_POLLING_INTERVAL: u64 = 500;
```
How was this chosen? I'm not against a hardcoded polling interval, but is there a reason to believe that `500` is the right number?

MortimerGoro replied:
Good question. There isn't a formula to decide the perfect value, but 500 seemed like a good balance between resource usage and desired precision. I saw similar values in native games and other implementations. In general these events (ie discovering new displays) don't need very high precision. We can reduce the value if needed because the pooling thread is only active when there is an active tab using WebVR APIs.

cvan replied:
can this constant be a pref?

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    fn register(&self) {
+        if let Some(wevbr_sender) = self.webvr_thread() {
```
Spelling is wrong.

emilio commented on components/script/dom/vr.rs:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    fn unregister(&self) {
+        if let Some(wevbr_sender) = self.webvr_thread() {
```
and here.

larsbergstrom commented:
@MortimerGoro Nice work - things are getting close! I think that @emilio and I have a few more issues to address but that it's almost there. There will certainly be some follow-up issues for the bigger items especially that @cvan mentioned, but they shouldn't block getting this landed pretty soon.

MortimerGoro commented:
@larsbergstrom @emilio PR updated!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14684) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
I moved the event polling interval value to a preference. Thanks @cvan for your suggestion.

Travis-ci failed with a non related problem: `No space left on device`

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14831) made this pull request unmergeable. Please resolve the merge conflicts.

larsbergstrom commented:
Looks good to me! Just rebase and r=me.

MortimerGoro commented:
@larsbergstrom rebase done ;)

jdm commented:
@bors-servo: r=larsbergstrom

bors-servo commented:
:pushpin: Commit 47ccd82 has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom 47ccd82ee45331b738029874f61df5b0f0f654c6 -->

bors-servo commented:
:hourglass: Testing commit 47ccd82 with merge d6d6655...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1752)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14874) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Sorry, that was #14323. We can keep working on merging it after a rebase.

MortimerGoro commented:
@jdm Ok, no problem. Rebase done and fixed proc_macro attribute no longer needed warnings with the latest rust compiler.

emilio commented:
@bors-servo r=larsbergstrom,emilio,jdm,nox,asajeffrey,cvan

bors-servo commented:
:pushpin: Commit c5705bf has been approved by `larsbergstrom,emilio,jdm,nox,asajeffrey,cvan`

<!-- @bors-servo r=larsbergstrom,emilio,jdm,nox,asajeffrey,cvan c5705bff5003675d6d266c698653834027a78227 -->

bors-servo commented:
:hourglass: Testing commit c5705bf with merge 47b776f...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/5115)

jdm commented:
@bors-servo: retry
* linker failed; might be disk space.

bors-servo commented:
:hourglass: Testing commit c5705bf with merge 518ef39...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4813), [arm32](http://build.servo.org/builders/arm32/builds/4977), [arm64](http://build.servo.org/builders/arm64/builds/5019), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5117), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1791), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1799), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5088), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5130), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1793), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1833), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/254), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/255)
